modules/artifact-activation/src/test/java/org/mule/runtime/module/artifact/activation/internal/classloader/DefaultArtifactClassLoaderResolverTestCase.java

```diff
@@ -0,0 +1,408 @@
+
+  public DefaultArtifactClassLoaderResolverTestCase() throws IOException {
+    muleHomeFolder = temporaryFolder.getRoot();
+  }
+
+  @Before
+  public void setup() {
+    artifactClassLoaderResolver = new DefaultArtifactClassLoaderResolver(moduleRepository, nativeLibraryFinderFactory);
+
+    plugin1Descriptor.setBundleDescriptor(PLUGIN1_BUNDLE_DESCRIPTOR);
+    plugin2Descriptor.setBundleDescriptor(PLUGIN2_BUNDLE_DESCRIPTOR);
+  }
+
+  @After
+  public void tearDown() {
+    deleteIfNeeded(getDomainsFolder());
+    deleteIfNeeded(new File(getMuleLibFolder(), "shared"));
+  }
+
+  private void deleteIfNeeded(File file) {
+    if (file.exists()) {
+      deleteQuietly(file);
+    }
+  }
+
+  @Test
+  public void createDomainDefaultClassLoader() {
+    DomainDescriptor descriptor = getTestDomainDescriptor(DEFAULT_DOMAIN_NAME);
+    final String artifactId = getDomainId(DEFAULT_DOMAIN_NAME);
+
+    final ArtifactClassLoader domainClassLoader =
+        artifactClassLoaderResolver.createDomainClassLoader(descriptor, artifactClassLoaderResolver::resolvePluginClassLoader);
+
+    assertThat(domainClassLoader.getClassLoader(), instanceOf(MuleSharedDomainClassLoader.class));
+    assertThat(domainClassLoader.getArtifactId(), equalTo(artifactId));
+  }
+
+  @Test
+  public void createDomainCustomClassLoader() {
+    final String domainName = "custom-domain";
+    DomainDescriptor descriptor = getTestDomainDescriptor(domainName);
+    final String artifactId = getDomainId(domainName);
+    descriptor.setRootFolder(createDomainDir(MULE_DOMAIN_FOLDER, domainName));
+
+    final ArtifactClassLoader domainClassLoader =
+        artifactClassLoaderResolver.createDomainClassLoader(descriptor, artifactClassLoaderResolver::resolvePluginClassLoader);
+
+    assertThat(domainClassLoader.getClassLoader(), instanceOf(MuleSharedDomainClassLoader.class));
+    assertThat(domainClassLoader.getArtifactId(), equalTo(artifactId));
+  }
+
+  @Test(expected = ArtifactActivationException.class)
+  public void validateDomainBeforeCreatingClassLoader() {
+    DomainDescriptor descriptor = getTestDomainDescriptor("some-domain");
+    descriptor.setRootFolder(new File("non-existent"));
+
+    artifactClassLoaderResolver.createDomainClassLoader(descriptor, artifactClassLoaderResolver::resolvePluginClassLoader);
+  }
+
+  @Test
+  public void createDomainClassLoaderWithExportedPackages() {
+    MuleModule muleModule = mock(MuleModule.class);
+    final String repeatedPackageName = "module&domain-package";
+    when(muleModule.getExportedPackages()).thenReturn(singleton(repeatedPackageName));
+    when(moduleRepository.getModules()).thenReturn(singletonList(muleModule));
+    final String onlyDomainPackageName = "domain-package";
+
+    final MuleDeployableArtifactClassLoader domainClassLoader =
+        getTestDomainClassLoader(emptySet(), Stream.of(onlyDomainPackageName, repeatedPackageName).collect(Collectors.toSet()));
```

Comment:
Static Import(only Collectors.toSet() because Stream.of collides with optional.of)